data/basal.go

```diff
@@ -0,0 +1,166 @@
+	Base         `bson:",inline"`
+}
+
+//SupressedBasal represents a suppressed basal portion of a basal
+type SupressedBasal struct {
+	Type         string  `json:"type" bson:"type" valid:"required"`
+	DeliveryType string  `json:"deliveryType" bson:"deliveryType" valid:"basaldeliverytype"`
+	ScheduleName string  `json:"scheduleName" bson:"scheduleName" valid:"omitempty,required"`
+	Rate         float64 `json:"rate" bson:"rate" valid:"omitempty,basalrate"`
+}
+
+const (
+	//BasalName is the given name for the type of a `Basal` datum
+	BasalName = "basal"
+
+	deliveryTypeField = "deliveryType"
+	scheduleNameField = "scheduleName"
+	insulinField      = "insulin"
+	valueField        = "value"
+	rateField         = "rate"
+	durationField     = "duration"
+
+	deliveryTypeTag validate.ValidationTag = "basaldeliverytype"
+	rateTag         validate.ValidationTag = "basalrate"
+	durationTag     validate.ValidationTag = "basalduration"
+	insulinTag      validate.ValidationTag = "basalinsulin"
+	valueTag        validate.ValidationTag = "basalvalue"
+
+	injectedDelivery, scheduledDelivery, suspendDelivery, tempDelivery = "injected", "scheduled", "suspend", "temp"
+
+	levemirInsulin, lantusInsulin = "levemir", "lantus"
+
+	rateValidationLowerLimit     = 0.0
+	durationValidationLowerLimit = 0
+	valueValidationLowerLimit    = 0
+	validationGreaterThanInt     = "Must be greater than %d"
+)
+
+var (
+	allowedDeliveryTypes = map[string]string{injectedDelivery: injectedDelivery, scheduledDelivery: scheduledDelivery, suspendDelivery: suspendDelivery, tempDelivery: tempDelivery}
+	allowedInsulins      = map[string]string{levemirInsulin: levemirInsulin, lantusInsulin: lantusInsulin}
+
+	basalFailureReasons = validate.ErrorReasons{
+		deliveryTypeTag: fmt.Sprintf("Must be one of %s,%s,%s,%s", injectedDelivery, scheduledDelivery, suspendDelivery, tempDelivery),
+		rateTag:         fmt.Sprintf("Must be greater than %f", rateValidationLowerLimit),
+		durationTag:     fmt.Sprintf(validationGreaterThanInt, durationValidationLowerLimit),
+		valueTag:        fmt.Sprintf(validationGreaterThanInt, valueValidationLowerLimit),
+		insulinTag:      fmt.Sprintf("Must be one of %s,%s", levemirInsulin, lantusInsulin),
+	}
+)
+
+//BuildBasal will build a Basal record
+func BuildBasal(obj map[string]interface{}) (*Basal, *Error) {
+
+	base, errs := BuildBase(obj)
+	cast := NewCaster(errs)
+
+	basal := &Basal{
+		DeliveryType: cast.ToString(deliveryTypeField, obj[deliveryTypeField]),
+		ScheduleName: cast.ToString(scheduleNameField, obj[scheduleNameField]),
+		Base:         base,
+	}
+
+	if obj[rateField] != nil {
+		basal.Rate = cast.ToFloat64(rateField, obj[rateField])
+	}
+	if obj[durationField] != nil {
+		basal.Duration = cast.ToInt(durationField, obj[durationField])
+	}
+	if obj[insulinField] != nil {
+		basal.Insulin = cast.ToString(insulinField, obj[insulinField])
+	}
+
+	if validationErrors := validator.Struct(basal); len(validationErrors) > 0 {
```

Comment:
Might want to only do validation if the previous casting hasn't returned any errors.  Could get a combination of confusing error messages.  You'll get an error about the casting failure and then a potential subsequent failure about the failed casted value (empty string, for example) not be valid either.